mmdet/datasets/coco.py

```diff
@@ -422,30 +425,34 @@ def evaluate(self,
         Returns:
             dict[str, float]: COCO style evaluation metric.
         """
-
-        metrics = metric if isinstance(metric, list) else [metric]
-        allowed_metrics = ['bbox', 'segm', 'proposal', 'proposal_fast']
-        for metric in metrics:
-            if metric not in allowed_metrics:
-                raise KeyError(f'metric {metric} is not supported')
         if iou_thrs is None:
             iou_thrs = np.linspace(
                 .5, 0.95, int(np.round((0.95 - .5) / .05)) + 1, endpoint=True)
         if metric_items is not None:
             if not isinstance(metric_items, list):
                 metric_items = [metric_items]
 
-        result_files, tmp_dir = self.format_results(results, jsonfile_prefix)
-
         eval_results = OrderedDict()
-        cocoGt = self.coco
+        if hasattr(self, 'THING_CLASSES') and hasattr(self, 'STUFF_CLASSES'):
+            cocoGt = COCO(self.ins_ann_file)
+            self.cat_ids = cocoGt.get_cat_ids(cat_names=self.THING_CLASSES)
+            assert self.ins_ann_file is not None, 'Annotation '\
+                'file for instance segmentation or object detection ' \
+                'shuold not be None'
+        else:
+            cocoGt = self.coco
+            self.cat_ids = cocoGt.get_cat_ids(cat_names=self.CLASSES)
```

Comment:
This logic is about distinguishing the class is CocoDataset or CocoPanopticDataset. It should not be placed in the CocoDataset. One way is to make `cocoGt` as an input of this function, CocoDataset and CocoPanopticDataset will pas different `cocoGt` into this function.

Reply:
Good idea.